openquake/hazardlib/shakemap/gmfs.py

```diff
@@ -151,41 +152,131 @@ def cholesky(spatial_cov, cross_corr):
     return numpy.linalg.cholesky(numpy.array(LLT))
 
 
-def to_gmfs(shakemap, spatialcorr, crosscorr, site_effects, trunclevel,
-            num_gmfs, seed, imts=None):
+gmfs_calculation_methods = CallableDict()
+
+MAXSITES = 1000
+CORRELATION_MATRIX_TOO_LARGE = '''\
+You have a correlation matrix which is too large: %s > %d.
+To avoid that, set a proper `region_grid_spacing` so that your exposure
+involves less sites.'''
+
+
+@gmfs_calculation_methods.add('Silva&Horspool')
+def gmfs_calculation_methods_SH(kind, shakemap, imts, spatialcorr,
+                                crosscorr, cholesky_limit):
     """
-    :returns: (IMT-strings, array of GMFs of shape (R, N, E, M)
+    Implementation of paper by Silva and Horspool 2019
+    https://onlinelibrary.wiley.com/doi/abs/10.1002/eqe.3154?af=R
+
+    :param shakemap: site coordinates with shakemap values
+    :param imts: list of required imts
+    :param spatialcorr: 'no', 'yes' or 'full'
+    :param crosscorr: 'no', 'yes' or 'full'
+    :returns: F(Z, mu) to calculate gmfs
     """
-    N = len(shakemap)  # number of sites
-    std = shakemap['std']
-    if imts is None or len(imts) == 0:
-        imts = std.dtype.names
-    else:
-        imts = [imt for imt in imts if imt in std.dtype.names]
-    val = {imt: numpy.log(shakemap['val'][imt]) for imt in imts}
-    imts_ = [imt.from_string(name) for name in imts]
-    M = len(imts_)
-    cross_corr = cross_correlation_matrix(imts_, crosscorr)
-    mu = numpy.array([numpy.ones(num_gmfs) * val[str(imt)][j]
-                      for imt in imts_ for j in range(N)])
+    # checks
+    N = len(shakemap)
+    M = len(imts)
+    if spatialcorr != 'no' and N > MAXSITES:
+        # hard-coded, heuristic
+        raise ValueError(CORRELATION_MATRIX_TOO_LARGE %
+                         (N, MAXSITES))
+    if N * M > cholesky_limit:
+        raise ValueError(CORRELATION_MATRIX_TOO_LARGE % (
+            '%d x %d' % (M, N), cholesky_limit))
```

Comment:
I understand the cholesky_limit check. 

But why did you add the check for the "MAXSITES" here, seeing that it is hardcoded. If it is for performance reasons when computing the spatial correlation, then the improvements from #6624 should have fixed this.

Reply:
It should not be hard-coded but configurable like the cholesky_limit. Consider it at technical debt.
It is not there for performance, but to avoid small negative eigenvalues due to numerical errors. It would be nicer to have a real solution rather than the ugly workaround :-(

Reply:
But why is only checking N*M not enough?

Reply:
Because in the formula numpy.linalg.cholesky(LLT) the matrix LLT has shape `(N*M, N*M)`

Reply:
I am sorry, I think I am a bit slow.
You are checking `if N*M > cholesky_limit:` - so why do you also check `if N > MAXSITES`?

If cholesky_limit = 10000 (default) and MAXSITES = 1000 (hardcoded) then:

N = 4000 and M = 1 => ValueError
N = 999 and M = 5 => no Value Error
even though N*M is bigger in the second example.

Reply:
The MAXSITES check was there before, introduced years ago, I just added the additional check. Perhaps we can fully remove the check on MAXSITES and instead add a check that there are no negative eigenvalues. Thanks to the cholesky_limit we know that the decomposition will still work without running out of memory.

Reply:
sorry I went a bit off-topic here. I resolve this conversation and leave you to accept / ask for additional changes in the pull request.